File: src/content/docs/topics/cs/coding-problems/arrays-and-hashing/347-top-k-frequent-elements.ts
function assert(condition: boolean, msg: string = ''): void {
    if (!condition) throw new Error(msg || 'Assertion failed');
}

function topKFrequent(nums: number[], k: number): number[] {
    // TODO: implement
    return [];
}

function _runTests(): void {
    assert(JSON.stringify([...topKFrequent([1, 1, 1, 2, 2, 3], 2)].sort((a, b) => a - b)) === JSON.stringify([1, 2]));
    assert(JSON.stringify(topKFrequent([1], 1)) === JSON.stringify([1]));
    assert(JSON.stringify([...topKFrequent([1, 2], 2)].sort((a, b) => a - b)) === JSON.stringify([1, 2]));
    const r = topKFrequent([1, 2, 3], 1);
    assert(r.length === 1 && [1, 2, 3].includes(r[0]));
    // perf
    const nums = Array.from({ length: 10000 }, (_, i) => i % 100);
    const t0 = performance.now();
    topKFrequent(nums, 10);
    console.log(`perf topKFrequent n=10000 k=10: ${(performance.now() - t0).toFixed(1)}ms`);
    console.log('all tests pass');
}

_runTests();

File: src/content/docs/topics/cs/coding-problems/arrays-and-hashing/347-top-k-frequent-elements.go
package main

import "fmt"

func assert(condition bool, msgs ...string) {
	if !condition {
		msg := "assertion failed"
		if len(msgs) > 0 {
			msg = msgs[0]
		}
		panic(msg)
	}
}

func topKFrequent(nums []int, k int) []int {
	// TODO: implement
	return nil
}

func sortedInts(s []int) []int {
	cp := make([]int, len(s))
	copy(cp, s)
	for i := 0; i < len(cp); i++ {
		for j := i + 1; j < len(cp); j++ {
			if cp[j] < cp[i] {
				cp[i], cp[j] = cp[j], cp[i]
			}
		}
	}
	return cp
}

func intsEqual(a, b []int) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}

func contains(s []int, v int) bool {
	for _, x := range s {
		if x == v {
			return true
		}
	}
	return false
}

func runTests() {
	assert(intsEqual(sortedInts(topKFrequent([]int{1, 1, 1, 2, 2, 3}, 2)), []int{1, 2}), "test 1")
	assert(intsEqual(topKFrequent([]int{1}, 1), []int{1}), "test 2")
	assert(intsEqual(sortedInts(topKFrequent([]int{1, 2}, 2)), []int{1, 2}), "test 3")
	r := topKFrequent([]int{1, 2, 3}, 1)
	assert(len(r) == 1 && contains([]int{1, 2, 3}, r[0]), "test 4")
	fmt.Println("all tests pass")
}

func main() { runTests() }
